### PR TITLE
Only do priming read if there is no space in dsts buffers.

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

A SSL_read is needed to ensure the bio buffer is flushed, for this we did a priming read. This can be removed in many cases. Also ensure we always fill as much as possible in the destination buffers.

Modifications:

- Only do priming read if capacity of all dsts buffers is zero
- Always produce as must data as possible in the dsts buffers.

Result:

Faster code.

### CODE REVIEW

**Comment by normanmaurer** [comment]
@trustin please review

File: handler/src/main/java/io/netty/handler/ssl/OpenSslEngine.java [hunk 1]
```diff
@@ -738,20 +696,20 @@ public synchronized SSLEngineResult unwrap(
...
                 }
 
-                if (pendingApp <= 0) {
-                    break;
-                }
-
                 int bytesRead = readPlaintextData(dst);
 
+                // TODO: We may want to consider if we move this check and only do it in a less often called place at
+                // the price to be not 100% accurate, like for example when calling SSL.getError(...).
```
Comment (trustin) [inline handler/src/main/java/io/netty/handler/ssl/OpenSslEngine.java]: at the price of not being 100% accurate

File: handler/src/main/java/io/netty/handler/ssl/OpenSslEngine.java [hunk 2]
```diff
@@ -772,9 +730,24 @@ public synchronized SSLEngineResult unwrap(
                     }
                 }
             }
+        } else {
+            // If the capacity of all destination buffers is 0 we need to trigger a SSL_read anyway to ensure
+            // everything is flushed in the BIO pair and so we can detect it in the pendingAppData() call.
+            if (SSL.readFromSSL(ssl, EMPTY_ADDR, 0) <= 0) {
+                // We not check SSL_get_error as we are not interested in any error that is not fatal.
```
Comment (trustin) [inline handler/src/main/java/io/netty/handler/ssl/OpenSslEngine.java]: We do not

**Comment by trustin** [comment]
LGTM. Fix the comments though.

File: handler/src/main/java/io/netty/handler/ssl/OpenSslEngine.java (hunk 1 above)
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/OpenSslEngine.java]: done

File: handler/src/main/java/io/netty/handler/ssl/OpenSslEngine.java (hunk 2 above)
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/OpenSslEngine.java]: done

**Comment by normanmaurer** [comment]
Cherry-picked into 4.0 (d345465bba7df9561584af0d939b22b81cc9ec64), 4.1 (8d1c6ebf71b116f9a33cce144a49e14f4e84528f) and master (477010e0162085202831baac8bc4d5ab78f51ca0)

**Pull request closed without merging**